### ISSUE TITLE
Can I use go1.21 to compile plugins?

### ISSUE THREAD

**Issue by inliquid** [knqyf263/go-plugin#58]
I just have tried to compile with go tool, since support for `wasip1` added in `go1.21`,  and got this error:
```
$ GOOS=wasip1 GOARCH=wasm go build -tags=tinygo.wasm -o wasm/plugin.wasm ./wasm
package gitlab.com/***/test-wasm-plugin/wasm
        imports gitlab.com/***/test-wasm-plugin/api
        imports github.com/knqyf263/go-plugin/wasm: build constraints exclude all Go files in C:\Users\***\go\pkg\mod\github.com\knqyf263\go-plugin@v0.8.0\wasm
```
Note that I'm passing `-tags=tinygo.wasm`, because my plugin code is also protected with this build constraint. However, error message sounds like if `go-plugin` is not intended for use with official go compiler at all?

**Comment by saschagrunert** [comment]
@knqyf263 do you see any chance to support cross compiling without tinygo?

**Comment by knqyf263** [comment]
I remember Go's wasip1 was complicated and difficult to run as Wasm, but it has been over a year since I tried it before, so we may try again. Contributions are welcome.

**Comment by saschagrunert** [comment]
The issue seems to be that we exclude this file because CGO is disabled for `GOOS=wasip1 GOARCH=wasm`:

https://github.com/knqyf263/go-plugin/blob/main/wasm/plugin.go

If we enable CGO
```
> CGO_ENABLED=1 GOOS=wasip1 GOARCH=wasm go build -tags=tinygo.wasm examples/helloworld/plugin-evening/evening.go
# runtime/cgo
cgo: unknown ptrSize for $GOARCH "wasm"
```

**Comment by knqyf263** [comment]
We need to figure out how to do the same thing with Go wasip1.
https://github.com/knqyf263/go-plugin/pull/45

**Comment by saschagrunert** [comment]
> We need to figure out how to do the same thing with Go wasip1. #45

I was playing around with the code a bit. Right now don't see any way to get the memory from a malloc without cgo. 

It also looks like that there will be no cgo support soon for the native golang wasm compiler: https://github.com/golang/go/issues/40543

**Comment by inliquid** [comment]
Big Go's wasip1 target still doesn't support exporting functions from Wasm module. See https://github.com/golang/go/issues/65199 So this means. until it's done it won't be feasible.

**Comment by knqyf263** [comment]
@inliquid Thanks for sharing the issue. This issue seems to be accepted and [the PR](https://go-review.googlesource.com/c/go/+/603055) has also been merged. Does it mean this export support will come in Go 1.24?

**Comment by inliquid** [comment]
@knqyf263 
> Thanks for sharing the issue. This issue seems to be accepted and [the PR](https://go-review.googlesource.com/c/go/+/603055) has also been merged. Does it mean this export support will come in Go 1.24?

Most likely it does. However there is another issue that needs to be considered. TinyGo does not actually stop the module when `main` is done as it should for "Command" type of wasip1 module (as opposed to "Reactor"). This makes it possible to call exported functions when main already exited, for instance:
```go
// main is required for TinyGo to compile to Wasm.
func main() {
	greeting.RegisterGreeter(GoodEvening{})
}
```
Exiting `main` normally should invalidate module and its exports. However, in practice it works differently, and there is [special note](https://github.com/tetratelabs/wazero/blob/main/RATIONALE.md#why-do-we-only-return-a-sysexiterror-on-a-non-zero-exit-code) in Wazero's docs:
> It is reasonable to think an exit error should be returned, even if the code is success (zero). Even on success, the module is no longer functional. For example, function exports would error later. However, wazero does not. The only time `sys.ExitError` is on error (non-zero).
>
> This decision was to improve performance and ergonomics for guests that both use WASI (have a `_start` function), and also allow custom exports. Specifically, Rust, TinyGo and normal wasi-libc, don't exit the module during `_start`. If they did, it would invalidate their function exports. This means it is unlikely most compilers will change this behavior.
> 
> `GOOS=waspi1` from Go 1.21 does exit during `_start`. However, it doesn't support other exports besides `_start`, and `_start` is not defined to be called multiple times anyway.

As far as I'm aware TinyGo merged "Reactor" support into `dev`, see https://github.com/tinygo-org/tinygo/pull/4451, but not sure when it's planned for release. 

Also, I'm not sure whether "Reactor" mode is 100% suitable for `go-plugin`. There are other things such as requirement to have exported linear memory to make host and guest able to communicate. Will it work the same way for "Reactors" module? I'm not sure at the moment.

In order to make it consistent with big Go's planned implementation I see following options:

1. Research and possibly switch to "Reactor" mode for both TinyGo and "big" Go
2. Another option could be to stay with "Command" type, but not exit `main` in case of "big" Go, but instead block on something like `nil` channel, like we have to do in case of `GOARCH=wasm GOOS=js`

**Comment by leonm1** [comment]
I recently implemented support for big Go for proxy-wasm as a wasi reactor in https://github.com/proxy-wasm/proxy-wasm-go-sdk/pull/1.

The required work included:

* Updating imports/exports to use the `//go:wasmimport` and `//go:wasmexport` directives
* Updating types at the ABI boundary to the limited types supported by the Go compiler (int32, int64, uint32, uint64, unsafe.Pointer, bool, string, and pointers to the allowed types)
* Adding new instructions for plugin authors to call global registration methods, since `main` is not called for a wasi reactor.
* compiling with `GOOS=wasip1 GOARCH=wasm go build -buildmode=c-shared`

It was blocked on the [wasmexport proposal](https://github.com/golang/go/issues/65199) and [WASI reactor support](https://go-review.googlesource.com/c/go/+/604316), which was implemented recently by the go core team.

> Also, I'm not sure whether "Reactor" mode is 100% suitable for go-plugin. There are other things such as requirement to have exported linear memory to make host and guest able to communicate. Will it work the same way for "Reactor" modules? I don't know at the moment.

I haven't looked deeply into `go-plugin`'s ABI, but a WASI reactor can import functions exported by the host (with the `//go:wasmimport` directive) and export functions for the host to call (via `//go:wasmexport`). 

The host is able to access the wasm module's linear memory, and pointers inside the linear memory can be alloc'd and passed by either side of the wasm module boundary.

I'd be happy to take a look or answer some questions if you have more questions about our approach!

**Comment by knqyf263** [comment]
According to @leonm1's explanation, adding big Go support doesn't sound so challenging. It would be great if someone could work on that.

**Comment by deadprogram** [comment]
I just became aware of the compatibility issues here with recent TinyGo releases.

Please see https://github.com/tinygo-org/tinygo/pull/4734 for one possible solution.

**Comment by knqyf263** [comment]
I confirmed Go 1.24 worked with go-plugin if we remove CGO.

**Comment by trading-peter** [comment]
Hi. I want to try go-plugin for the first time and came across this issue.

> I confirmed Go 1.24 worked with go-plugin if we remove CGO.

Does that mean one can simply use Go 1.24 to build a plugin as long as CGO is disabled or does this statement mean that its now doable but needs internal adjustments in the code of go-plugin first?

**Comment by saschagrunert** [comment]
I guess we need to find a cgo-free implementation of https://github.com/knqyf263/go-plugin/blob/main/wasm/plugin.go#L5-L35 or work around the whole conversion at all.

**Comment by leonm1** [comment]
It's a bit hard for me to mentally parse the codegen without any prior experience with this project, but it seems like the additional malloc and free aren't unnecessary.

In Proxy-Wasm for example, when the host environment needs to copy data into wasm, the interaction works like so:

1. The wasm code will call a hostcall which requires allocation. In proxy-wasm, this would be something like `proxy_get_buffer_bytes`, which allocates and writes the request body into the wasm address space.
2. The host will call the exported wasm function [`proxy_on_memory_allocate(size uint32)`](https://github.com/proxy-wasm/proxy-wasm-go-sdk/blob/ab4161dcf9246a828008b539a82a1556cf0f2e24/proxywasm/internal/abi_callback_alloc.go), which allocates a byte slice of the desired size and returns a pointer to the beginning of it.
3. The host writes data to that address (using the utilities provided by the wasm runtime being used)
4. The host function called in (1) returns, passing the pointer received from `proxy_on_memory_allocate`
5. The wasm code calls `unsafe.Slice(ptr, size)` to turn the raw pointer into a (safe) byte slice again.

I'm no expert, but as far as I understand the way the Go runtime is returning from all these wasm <-> host calls, the garbage collector won't free the memory as long as there are no allocations before using unsafe.Slice in step (5).

If you'd prefer a stronger guarantee, you could keep the byte slices alloc'd byte slices from `ByteToPtr` in a global struct / map to keep a GC reference alive until it is manually free'd in `FreePtr`.

**Comment by inliquid** [comment]
@leonm1 
> The wasm code will call a hostcall which requires allocation. In proxy-wasm, this would be something like proxy_get_buffer_bytes, which allocates and writes the request body into the wasm address space.

With plugins it's Host who initiates a call to Guest and needs to allocate before control is passed to Wasm code. Does it matter in the context of logic of Proxy-Wasm?

The `malloc`/`free` are generated by TinyGo and exported by the Core Module, as well as its Memory, so that Host can call them (it's in generated code):
```go
        <...>
	malloc := module.ExportedFunction("malloc")
	if malloc == nil {
		return nil, errors.New("malloc is not exported")
	}

	free := module.ExportedFunction("free")
	if free == nil {
		return nil, errors.New("free is not exported")
	}
        <...>
```

So here is what's in the module:
```
<...>
 (export "memory" (memory $0))
 (export "malloc" (func $malloc))
 (export "free" (func $free))
 (export "calloc" (func $calloc))
 (export "realloc" (func $realloc))
 (export "_start" (func $_start))
<...>
```

It's then used in the generated code like this:
```go
        var dataPtr uint64
	// If the input data is not empty, we must allocate the in-Wasm memory to store it, and pass to the plugin.
	if dataSize != 0 {
		results, err := p.malloc.Call(ctx, dataSize)
		if err != nil {
			return nil, err
		}
		dataPtr = results[0]
		// This pointer is managed by TinyGo, but TinyGo is unaware of external usage.
		// So, we have to free it when finished
		defer p.free.Call(ctx, dataPtr)

		// The pointer is a linear memory offset, which is where we write the name.
		if !p.module.Memory().Write(uint32(dataPtr), data) {
			return nil, fmt.Errorf("Memory.Write(%d, %d) out of range of memory size %d", dataPtr, dataSize, p.module.Memory().Size())
		}
	}
```

**Comment by knqyf263** [comment]
I implemented a simple memory allocator in Go and exported malloc and free. Based on my tests, there were no memory leaks, and it seemed to work correctly. It would be helpful if someone else could test it, but since it appears to be functioning properly in my verification, I'll merge it.
https://github.com/knqyf263/go-plugin/pull/71

**Comment by knqyf263** [comment]
I don’t think it’s difficult to continue supporting both Go and TinyGo. In fact, I have kept the files for TinyGo.
https://github.com/knqyf263/go-plugin/blob/main/wasm/plugin_tinygo.go#L1

However, I’m not sure if I can maintain both in my spare time, so for now, I’m focusing on Go support. It would be great if someone could take over the maintenance of TinyGo support in go-plugin.

**Comment by deadprogram** [comment]
At least as of TinyGo 0.36+ the compatibility for compilation should be a lot better, which should hopefully remove most of the effort required.

**Comment by knqyf263** [comment]
Yes, I believe go-plugin works with TinyGo v0.36+. I just didn't have time to test it. Please let me know if someone tests it and ensures that.

**Comment by knqyf263** [comment]
It's out
https://github.com/knqyf263/go-plugin/releases/tag/v0.9.0
